src-plugins/mscExportVideo/mscExportVideo.cpp

```diff
@@ -0,0 +1,317 @@
+#include "mscExportVideo.h"
+
+#include <dtkCore/dtkAbstractProcessFactory.h>
+
+#include <vtkImageCanvasSource2D.h>
+#include <vtkImageData.h>
+#include <vtkJPEGWriter.h>
+#include <vtkOggTheoraWriter.h>
+#include <vtkQImageToImageSource.h>
+#include <vtkSmartPointer.h>
+
+// /////////////////////////////////////////////////////////////////
+// ExportVideoPrivate
+// /////////////////////////////////////////////////////////////////
+namespace msc
+{
+class ExportVideoPrivate
+{
+public:
+    QVector< QImage > imagesArray;
+    QVector< QColor > tmpColorArray;
+    int tmpRed;
+    int tmpGreen;
+    int width;
+    int height;
+
+    // User parameters
+    QString filename;
+    int format;            // Cf. VideoFormat in header file
+    int frameRate;         // Frame per second
+    bool subsampling;      // Is the video to be encoded using 4:2:0 subsampling?
+    int quality;           // 0 = low, 1 = medium, 2 = high
+    bool exportOnlyAsJPEG; // Can be used for home encoding
+};
+
+// /////////////////////////////////////////////////////////////////
+// ExportVideo
+// /////////////////////////////////////////////////////////////////
+
+ExportVideo::ExportVideo() : medAbstractProcess(), d(new ExportVideoPrivate)
+{
+    d->imagesArray.clear();
+    d->tmpColorArray.clear();
+    d->tmpRed = 0;
+    d->tmpGreen = 0;
+    d->width  = 0;
+    d->height = 0;
+
+    // User parameters
+    d->format = OGGVORBIS;
+    d->filename = "";
+    d->frameRate = 1;
+    d->subsampling = false;
+    d->quality = 2;
+    d->exportOnlyAsJPEG = false;
+}
+
+ExportVideo::~ExportVideo()
+{
+    delete d;
+}
+
+bool ExportVideo::registered()
+{
+    return dtkAbstractProcessFactory::instance()->registerProcessType("ExportVideo", createExportVideo);
+}
+
+QString ExportVideo::description() const
+{
+    return "ExportVideo";
+}
+
+QString ExportVideo::identifier() const
+{
+    return description();
+}
+
+void ExportVideo::setParameter(int data, int channel)
```

Comment:
(I'm sorry if the following is not clear, but there really is a serious design issue here)

This uses ```setParameter``` in a way it was not intended for:
- For each pixel, you must set channels 0 and 1 first, and then channel 2, and repeat this for all pixels.
- You must set channel 4 after having set channel 3 and after having passed all pixel values through channels 0, 1 and 2.
- The previous calls must be repeated for each frame of the video, in the order of the frames.

The ```setParameter``` function should not be required to be called in a specific order, nor should it be called  multiple times for one parameter.
There are many ways to avoid this. For example, you could do the following:
- Use one channel to set all pixel values of one frame, in one single call using ```setParameter(int* data, int channel, int frame)```. The size of the array could be stored as the first value of the array (since DTK strangely does not provide a ```count``` parameter as it does for a double array), and the rest of the pixel values would be stored in consecutive order R, G, B, R, G B etc. The pixels would be stored in a ```QVector<QVector<QColor>>``` (each element of this array being one frame of colors). To avoid forcing the caller to send each frame in order (0, 1, 2 etc.) you would simply store the array directly where it belongs using ```QVector::insert``` (To make sure the vector is big enough, you can resize it if needed).
- The ```QImage``` created in case 4 should be created only when running the filter (before the call to ```vtkQImageToImageSource::SetQImage```), so that you do not have to set the colors before the size of the image (as a general case, any data the depends on multiple parameters should not be created in a ```setParameter``` call).